var/spack/repos/builtin/packages/ucx/package.py

```diff
@@ -43,14 +45,15 @@ class Ucx(AutotoolsPackage, CudaPackage):
     variant('assertions', default=False,
             description='Enable assertions')
     variant('parameter_checking', default=False,
-            description='Enable paramter checking')
+            description='Enable parameter checking')
     variant('pic', default=True,
             description='Builds with PIC support')
     variant('java', default=False,
             description='Builds with Java bindings')
     variant('gdrcopy', default=False,
             description='Enable gdrcopy support')
 
+    depends_on('binutils')
```

Comment:
Can you provide a bit more insight on why this dependency is necessary? Is it to compile it on old operating systems?

Reply:
When I was building on a compute node without binutils (`libbfd`) in the image, I was unable to compile ucx because during configure it fails with `Please use compiler that supports __attribute__((constructor))`, but this is because that check couldn't find `libbfd`.

Reply:
This was on a CentOS 7 system btw.

Reply:
I'm opposed to adding this dependency for the reason documented at https://spack.readthedocs.io/en/latest/getting_started.html#binutils

Basically, almost every compiled package in Spack depends on binutils, but we don't want to add that dependency to every package. When I run into these issues, I build binutils, run `spack load binutils`, and then `spack install ucx`.

Reply:
I tend to agree with @adamjstewart Can we remove the dependency from the PR?

Reply:
I would like to agree with you both. However, I always have binutils loaded. The problem here is that Spack sanitizes the `LD_LIBRARY_PATH` right? That's why this didn't work for me without the explicit dependency.

Reply:
Spack sanitizes `LD_LIBRARY_PATH`, but not `PATH`. In my experience, it's usually the executables that come with `binutils` that are needed, not the libraries.

Reply:
That's the problem with UCX though. It is linking to `libbfd` during configure to test that the compiler meets its requirements.

Reply:
I removed the dependency because I can workaround it, but I disagree with doing so because not having it caused me a fair amount of confusion and research to understand why compiler didn't support ` __attribute__((constructor))`, when it was looking for a library from `binutils`.

Reply:
@jrood-nrel Two questions to see if we can find a solution for you that is not a workaround:

1. Can you post the entry of your compiler in `compilers.yaml`?
2. What do you need set in `LD_LIBRARY_PATH`? If you compiler with `--dirty` is the compilation working?

Reply:
Here is the compiler I was using when I experienced this https://github.com/jrood-nrel/spack-configs/blob/02ffaca1930e923e5ca554802211332cbb464a9a/configs/eagle/software/compilers.yaml#L62-L73 . `libbfd` resides in `/yadda/yadda/spack/opt/spack/linux-centos7-skylake_avx512/gcc-8.4.0/binutils-2.34-2drppsfmkeudqr3ll7zz7pqxlsszuh4b/lib` so I would need that in the `LD_LIBRARY_PATH`. I haven't tried `--dirty`, but that wouldn't work for my case since `libbfd.so` is not in the compute node image I was compiling on.

Reply:
> I haven't tried --dirty, but that wouldn't work for my case since libbfd.so is not in the compute node image I was compiling on.

I don't fully understand this last sentence. If you need a path added to `LD_LIBRARY_PATH` and you use `--dirty` that would prevent unsetting `LD_LIBRARY_PATH` when Spack is compiling. Anyhow, I would still consider this a workaround.

A better solution if you need environment modifications related to a compiler should be to use the `env` attribute:
```yaml
  compiler:
    extra_rpaths:
     - /nopt/nrel/ecom/hpacf/compilers/2020-07/spack/opt/spack/linux-centos7-skylake_avx512/gcc-8.4.0/gcc-8.4.0-2a3vha6hlw4xc5ja3jyhr7huzaxuw2kt/lib64
    modules: []
    operating_system: centos7
    paths:
      cc:  /nopt/nrel/ecom/hpacf/compilers/2020-07/spack/opt/spack/linux-centos7-skylake_avx512/gcc-8.4.0/gcc-8.4.0-2a3vha6hlw4xc5ja3jyhr7huzaxuw2kt/bin/gcc
      cxx: /nopt/nrel/ecom/hpacf/compilers/2020-07/spack/opt/spack/linux-centos7-skylake_avx512/gcc-8.4.0/gcc-8.4.0-2a3vha6hlw4xc5ja3jyhr7huzaxuw2kt/bin/g++
      f77: /nopt/nrel/ecom/hpacf/compilers/2020-07/spack/opt/spack/linux-centos7-skylake_avx512/gcc-8.4.0/gcc-8.4.0-2a3vha6hlw4xc5ja3jyhr7huzaxuw2kt/bin/gfortran
      fc:  /nopt/nrel/ecom/hpacf/compilers/2020-07/spack/opt/spack/linux-centos7-skylake_avx512/gcc-8.4.0/gcc-8.4.0-2a3vha6hlw4xc5ja3jyhr7huzaxuw2kt/bin/gfortran
    environment:
      prepend_path:
        LD_LIBRARY_PATH: '/yadda/yadda/spack/opt/spack/linux-centos7-skylake_avx512/gcc-8.4.0/binutils-2.34-2drppsfmkeudqr3ll7zz7pqxlsszuh4b/lib'
    spec: gcc@8.4.0
    target: x86_64
```
Can you check if this works in your case?

Reply:
Ok, I see what you mean as a workaround if I installed `binutils` manually with Spack and then added it to `LD_LIBRARY_PATH`. I was coming from the standpoint of not using `binutils` in Spack at all. I agree this is still just a workaround. Personally I find it easier as a workaround to have my own custom `ucx` with `depends_on('binutils')` 🙂 After reading that note in the documentation about binutils, I don't see why having every package depend on binutils would be a problem. Most packages either depend on CMake or Autotools. I'm not sure why `binutils` is any different. Seems like it defeats the purpose of Spack. Anyway, thanks for merging this.

Reply:
I guess one last note is that I would never plan on using a solution which involves using `--dirty`.